File: 735.行星碰撞.py
#
# @lc app=leetcode.cn id=735 lang=python3
#
# [735] 行星碰撞
#

# @lc code=start
from typing import List
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        s = []
        for i in asteroids:
            if not s: s.append(i)
            else:
                if not (s[-1] > 0 and i < 0):
                    s.append(i)
                    continue
                if s[-1] > -i: continue
                elif s[-1] < -i:
                    s.pop()
                    # goto line13
                    s.append(i)
                else: s.pop()
        return s
# @lc code=end

print(Solution().asteroidCollision(asteroids = [5,10,-5]))
print(Solution().asteroidCollision(asteroids = [8,-8]))

File: 735.行星碰撞.cpp
/*
 * @lc app=leetcode.cn id=735 lang=cpp
 *
 * [735] 行星碰撞
 */

// @lc code=start
#include<vector>
using namespace std;
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        vector<int> s;
        for(auto i: asteroids) {
            mylabel:
            if(s.empty()) s.push_back(i);
            else{
                if(!(s.back() > 0 && i < 0)){
                    s.push_back(i);
                    continue;
                }
                if(s.back() > -i) continue;
                else if (s.back() < -i){
                    s.pop_back();
                    goto mylabel;
                }
                else s.pop_back();
            }
        }
        return s;
    }
};
// @lc code=end

